search_knn.go

```diff
@@ -293,18 +289,21 @@ func mergeKNN(req *SearchRequest, sr *SearchResult) {
 	sr.Total = uint64(len(nonZeroScoreHits))
 }
 
-func recomputeTotalScore(operator int, hit *search.DocumentMatch) float64 {
+func recomputeTotalScore(operator knnOperator, hit *search.DocumentMatch) float64 {
 	totalScore := 0.0
-	numNonZero := 0
+	numNonZero := 0.0
+	numTotal := float64(len(hit.ScoreBreakdown))
 	for _, score := range hit.ScoreBreakdown {
 		if score != 0 {
 			numNonZero += 1
 		}
 		totalScore += score
 	}
-	if operator == 0 {
-		coord := float64(numNonZero) / float64(len(hit.ScoreBreakdown))
+	if operator == knnOperatorOr || operator == "" {
+		coord := numNonZero / numTotal
 		totalScore = totalScore * coord
+	} else if operator == knnOperatorAnd && numNonZero != numTotal {
+		totalScore = 0
```

Comment:
Why should totalScore be 0 in this case?

Reply:
that is because some of the scores of the subsearchers of the conjunction searcher have become zero.

for example - you have match + knn query(k=2) with operator(and)

with 2 partitions, you can get 4 results; their score breakdown can be like this
D1 - [1,2]
D2 - [2,3]
D3 - [3,4]
D4 - [4,5] (index 0 = match score and index 1 = knn score)

so you choose top 2 = D3 and D4 in the heap. => numNonZero = numTotal = 2
the other two D1 and D2 will be made [1,0] and [2,0] (following the logic if disjunction (operator or) was used) -> now D1 and D2 essentially satisfy only match query which breaks conjunction clause set earlier in operator. So score is zero => remove them from hits